{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMe2NIEY/O1tiU12xx0CTgc",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/VikramanDeepsphere/AI_Classifier/blob/main/Deepsphere-AI/UCD/UCD.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### ProteinParams"
      ],
      "metadata": {
        "id": "pYxydaacfLP6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "This program calculates the physical-chemical properties of a protein sequence\n",
        "\n",
        "input: a protein sequence\n",
        "    - as a one - letter encoding of the amino acid\n",
        "    - a valid amino is a capital letter and is in the table\n",
        "    - anything that violates this will not be stored in the protein string\n",
        "output: number of amino acids and total molecular weight\n",
        "       Molar extinction coefficient and Mass extinction coefficient\n",
        "       theoretical isoelectric point(pI)\n",
        "       amino acid composition\n",
        "\"\"\"\n",
        "\n",
        "class ProteinParam :\n",
        "# These tables are for calculating:\n",
        "#     molecular weight (aa2mw), along with the mol. weight of H2O (mwH2O)\n",
        "#     absorbance at 280 nm (aa2abs280)\n",
        "#     pKa of positively charged Amino Acids (aa2chargePos)\n",
        "#     pKa of negatively charged Amino acids (aa2chargeNeg)\n",
        "#     and the constants aaNterm and aaCterm for pKa of the respective termini\n",
        "#  Feel free to move these to appropriate methods as you like\n",
        "\n",
        "# As written, these are accessed as class attributes, for example:\n",
        "# ProteinParam.aa2mw['A'] or ProteinParam.mwH2O\n",
        "\n",
        "    aa2mw = {\n",
        "    'A': 89.093,  'G': 75.067,  'M': 149.211, 'S': 105.093, 'C': 121.158,\n",
        "    'H': 155.155, 'N': 132.118, 'T': 119.119, 'D': 133.103, 'I': 131.173,\n",
        "    'P': 115.131, 'V': 117.146, 'E': 147.129, 'K': 146.188, 'Q': 146.145,\n",
        "    'W': 204.225,  'F': 165.189, 'L': 131.173, 'R': 174.201, 'Y': 181.189\n",
        "    }\n",
        "    mwH2O = 18.015\n",
        "    aa2abs280= {'Y':1490, 'W': 5500, 'C': 125}\n",
        "    aa2chargeNeg = {'D': 3.86, 'E': 4.25, 'C': 8.33, 'Y': 10}\n",
        "    aaNterm = 9.69\n",
        "    aaCterm = 2.34\n",
        "    aa2chargePos = {'K': 10.5, 'R':12.4, 'H':6}\n",
        "\n",
        "    # the __init__ method requires a protein string to be provided, either as a\n",
        "    # string or list of strings that will be concatenated\n",
        "\n",
        "    \"\"\"\n",
        "    Here I loop through all the characters in the protein and and check if\n",
        "    they are allowed input, if so add to the splitAminos array. Then create\n",
        "    the protein string\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__ (self, protein):\n",
        "        splitAminos = []\n",
        "        allowed_aminos = self.aa2mw.keys()\n",
        "        for char in protein:\n",
        "            if char in allowed_aminos:\n",
        "                splitAminos.append(char)\n",
        "\n",
        "        l = ''.join(splitAminos).split()\n",
        "        self.protString = ''.join(l).upper()\n",
        "\n",
        "        print(self.protString)\n",
        "\n",
        "\n",
        "#Here the Count of the protein is returned\n",
        "    def aaCount (self):\n",
        "        return len(self.protString)\n",
        "\n",
        "#Here the PI is calculated by finding the ph at which the net charge is closest to 0 and non-negative\n",
        "    def pI (self):\n",
        "        ph = 0.00\n",
        "        lowestCharge = (99999,0.0)\n",
        "        while(ph <= 14.0):\n",
        "            currentCharge = self._charge_(ph)\n",
        "            if currentCharge < lowestCharge[0] and self._charge_(ph) >=0:\n",
        "                lowestCharge = (currentCharge,ph)\n",
        "            ph += 0.01\n",
        "\n",
        "        return lowestCharge[1]\n",
        "\n",
        "    \"\"\"\n",
        "    Here the dictionary of amino acids to their count is created and returned\n",
        "    \"\"\"\n",
        "    def aaComposition (self) :\n",
        "        compDict = {}\n",
        "        for amino in self.aa2mw.keys():\n",
        "            compDict[amino] = self.protString.count(amino)\n",
        "        return compDict\n",
        "\n",
        "    \"\"\"\n",
        "    Here the charge is calculated via the formula provided. Moreover, the acutual work is\n",
        "    done in the dotPh\n",
        "    \"\"\"\n",
        "    def _charge_ (self, pH):\n",
        "        firstCharge = self.dotPh(['R','K','H'],pH) + (10**self.aaNterm)/(10**self.aaNterm+10**pH)\n",
        "        secondCharge = self.dotPh(['D','E','C','Y'],pH) + (10**pH)/(10**self.aaCterm+10**pH)\n",
        "        return firstCharge - secondCharge\n",
        "    \"\"\"\n",
        "    Here the inner sum of each of the aminos is calculated\n",
        "    \"\"\"\n",
        "    def dotPh(self,allowed_aminos,pH):\n",
        "        totalFirstCharge = 0\n",
        "        for amino in self.protString:\n",
        "            if amino in allowed_aminos:\n",
        "                if amino in self.aa2chargePos.keys():\n",
        "                    t = 10**self.aa2chargePos[amino]\n",
        "                    b = 10**self.aa2chargePos[amino] + 10**pH\n",
        "                    totalFirstCharge += t/b\n",
        "                elif amino in self.aa2chargeNeg.keys():\n",
        "\n",
        "                    t = 10**pH\n",
        "                    b = 10**self.aa2chargeNeg[amino] + 10**pH\n",
        "                    totalFirstCharge += t/b\n",
        "                else:\n",
        "                    t = 10**0\n",
        "                    b = 10**0 + 10**pH\n",
        "                    totalFirstCharge += t/b\n",
        "        return totalFirstCharge\n",
        "\n",
        "    \"\"\"\n",
        "    Here the associated count of the aminos is multiplied by the associated constant\n",
        "    according to the formula provided.\n",
        "    \"\"\"\n",
        "    def molarExtinction (self):\n",
        "       return float(self.protString.count('Y')*self.aa2abs280['Y']\n",
        "        + self.protString.count('W')*self.aa2abs280['W']\n",
        "        + self.protString.count('C')*self.aa2abs280['C'])\n",
        "\n",
        "    def massExtinction (self):\n",
        "        myMW =  self.molecularWeight()\n",
        "        return self.molarExtinction() / myMW if myMW else 0.0\n",
        "\n",
        "    \"\"\"\n",
        "    Here the total molecularWeight is calculated by looping\n",
        "    through the sum of the molecularWeights associated to the amino acid\n",
        "    and then subtracting the waters for the peptide bonds.\n",
        "    As can be seen in the fomula.\n",
        "    \"\"\"\n",
        "    def molecularWeight (self):\n",
        "        firstMolecWeight = 0.0\n",
        "        length = sum(len(s) for s in self.protString)\n",
        "\n",
        "        for amino in self.protString:\n",
        "            firstMolecWeight += self.aa2mw[amino]\n",
        "        return (firstMolecWeight - self.mwH2O * (length-1))\n",
        "\n",
        "\n",
        "# Please do not modify any of the following.  This will produce a standard output that can be parsed\n",
        "\n",
        "import sys\n",
        "for inString in sys.stdin :\n",
        "    myParamMaker = ProteinParam(inString)\n",
        "    myAAnumber = myParamMaker.aaCount()\n",
        "    print (\"Number of Amino Acids: {aaNum}\".format(aaNum = myAAnumber))\n",
        "    print (\"Molecular Weight: {:.1f}\".format(myParamMaker.molecularWeight()))\n",
        "    print (\"molar Extinction coefficient: {:.2f}\".format(myParamMaker.molarExtinction()))\n",
        "    print (\"mass Extinction coefficient: {:.2f}\".format(myParamMaker.massExtinction()))\n",
        "    print (\"Theoretical pI: {:.2f}\".format(myParamMaker.pI()))\n",
        "    print (\"Amino acid composition:\")\n",
        "    myAAcomposition = myParamMaker.aaComposition()\n",
        "    keys = list(myAAcomposition.keys())\n",
        "    keys.sort()\n",
        "    if myAAnumber == 0 : myAAnumber = 1  # handles the case where no AA are present\n",
        "    for key in keys :\n",
        "        print (\"\\t{} = {:.2%}\".format(key, myAAcomposition[key]/myAAnumber))"
      ],
      "metadata": {
        "id": "xRA5HjwReaJd"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### SequenceAnalysis"
      ],
      "metadata": {
        "id": "DfUXlW0Wfyuf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class ProteinParam :\n",
        "# These tables are for calculating:\n",
        "#     molecular weight (aa2mw), along with the mol. weight of H2O (mwH2O)\n",
        "#     absorbance at 280 nm (aa2abs280)\n",
        "#     pKa of positively charged Amino Acids (aa2chargePos)\n",
        "#     pKa of negatively charged Amino acids (aa2chargeNeg)\n",
        "#     and the constants aaNterm and aaCterm for pKa of the respective termini\n",
        "#  Feel free to move these to appropriate methods as you like\n",
        "\n",
        "# As written, these are accessed as class attributes, for example:\n",
        "# ProteinParam.aa2mw['A'] or ProteinParam.mwH2O\n",
        "\n",
        "    aa2mw = {\n",
        "    'A': 89.093,  'G': 75.067,  'M': 149.211, 'S': 105.093, 'C': 121.158,\n",
        "    'H': 155.155, 'N': 132.118, 'T': 119.119, 'D': 133.103, 'I': 131.173,\n",
        "    'P': 115.131, 'V': 117.146, 'E': 147.129, 'K': 146.188, 'Q': 146.145,\n",
        "    'W': 204.225,  'F': 165.189, 'L': 131.173, 'R': 174.201, 'Y': 181.189\n",
        "    }\n",
        "    mwH2O = 18.015\n",
        "    aa2abs280= {'Y':1490, 'W': 5500, 'C': 125}\n",
        "    aa2chargeNeg = {'D': 3.86, 'E': 4.25, 'C': 8.33, 'Y': 10}\n",
        "    aaNterm = 9.69\n",
        "    aaCterm = 2.34\n",
        "    aa2chargePos = {'K': 10.5, 'R':12.4, 'H':6}\n",
        "\n",
        "    # the __init__ method requires a protein string to be provided, either as a\n",
        "    # string or list of strings that will be concatenated\n",
        "\n",
        "    \"\"\"\n",
        "    Here I loop through all the characters in the protein and and check if\n",
        "    they are allowed input, if so add to the splitAminos array. Then create\n",
        "    the protein string\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__ (self, protein):\n",
        "        splitAminos = []\n",
        "        allowed_aminos = self.aa2mw.keys()\n",
        "        for char in protein:\n",
        "            if char in allowed_aminos:\n",
        "                splitAminos.append(char)\n",
        "\n",
        "        l = ''.join(splitAminos).split()\n",
        "        self.protString = ''.join(l).upper()\n",
        "\n",
        "\n",
        "\n",
        "#Here the Count of the protein is returned\n",
        "    def aaCount (self):\n",
        "        return len(self.protString)\n",
        "\n",
        "#Here the PI is calculated by finding the ph at which the net charge is closest to 0 and non-negative\n",
        "    def pI (self):\n",
        "        ph = 0.00\n",
        "        lowestCharge = (99999,0.0)\n",
        "        while(ph <= 14.0):\n",
        "            currentCharge = self._charge_(ph)\n",
        "            if currentCharge < lowestCharge[0] and self._charge_(ph) >=0:\n",
        "                lowestCharge = (currentCharge,ph)\n",
        "            ph += 0.01\n",
        "\n",
        "        return lowestCharge[1]\n",
        "\n",
        "    \"\"\"\n",
        "    Here the dictionary of amino acids to their count is created and returned\n",
        "    \"\"\"\n",
        "    def aaComposition (self) :\n",
        "        compDict = {}\n",
        "        for amino in self.aa2mw.keys():\n",
        "            compDict[amino] = self.protString.count(amino)\n",
        "        return compDict\n",
        "\n",
        "    \"\"\"\n",
        "    Here the charge is calculated via the formula provided. Moreover, the acutual work is\n",
        "    done in the dotPh\n",
        "    \"\"\"\n",
        "    def _charge_ (self, pH):\n",
        "        firstCharge = self.dotPh(['R','K','H'],pH) + (10**self.aaNterm)/(10**self.aaNterm+10**pH)\n",
        "        secondCharge = self.dotPh(['D','E','C','Y'],pH) + (10**pH)/(10**self.aaCterm+10**pH)\n",
        "        return firstCharge - secondCharge\n",
        "    \"\"\"\n",
        "    Here the inner sum of each of the aminos is calculated\n",
        "    \"\"\"\n",
        "    def dotPh(self,allowed_aminos,pH):\n",
        "        totalFirstCharge = 0\n",
        "        for amino in self.protString:\n",
        "            if amino in allowed_aminos:\n",
        "                if amino in self.aa2chargePos.keys():\n",
        "                    t = 10**self.aa2chargePos[amino]\n",
        "                    b = 10**self.aa2chargePos[amino] + 10**pH\n",
        "                    totalFirstCharge += t/b\n",
        "                elif amino in self.aa2chargeNeg.keys():\n",
        "\n",
        "                    t = 10**pH\n",
        "                    b = 10**self.aa2chargeNeg[amino] + 10**pH\n",
        "                    totalFirstCharge += t/b\n",
        "                else:\n",
        "                    t = 10**0\n",
        "                    b = 10**0 + 10**pH\n",
        "                    totalFirstCharge += t/b\n",
        "        return totalFirstCharge\n",
        "\n",
        "    \"\"\"\n",
        "    Here the associated count of the aminos is multiplied by the associated constant\n",
        "    according to the formula provided.\n",
        "    \"\"\"\n",
        "    def molarExtinction (self):\n",
        "       return float(self.protString.count('Y')*self.aa2abs280['Y']\n",
        "        + self.protString.count('W')*self.aa2abs280['W']\n",
        "        + self.protString.count('C')*self.aa2abs280['C'])\n",
        "\n",
        "    def massExtinction (self):\n",
        "        myMW =  self.molecularWeight()\n",
        "        return self.molarExtinction() / myMW if myMW else 0.0\n",
        "\n",
        "    \"\"\"\n",
        "    Here the total molecularWeight is calculated by looping\n",
        "    through the sum of the molecularWeights associated to the amino acid\n",
        "    and then subtracting the waters for the peptide bonds.\n",
        "    As can be seen in the fomula.\n",
        "    \"\"\"\n",
        "    def molecularWeight (self):\n",
        "        firstMolecWeight = 0.0\n",
        "        length = sum(len(s) for s in self.protString)\n",
        "\n",
        "        for amino in self.protString:\n",
        "            firstMolecWeight += self.aa2mw[amino]\n",
        "        return (firstMolecWeight - self.mwH2O * (length-1))\n",
        "\n",
        "class NucParams:\n",
        "    \"\"\"\n",
        "    This class handles reading and processing data about a given genome sequence\n",
        "\n",
        "    methods:\n",
        "    __init__ (self)\n",
        "    addSequence (self, thisSequence)\n",
        "    aaComposition(self)\n",
        "    nucComposition(self)\n",
        "    codonComposition(self)\n",
        "    nucCount(self)\n",
        "\n",
        "    \"\"\"\n",
        "    rnaCodonTable = {\n",
        "    # RNA codon table\n",
        "    # U\n",
        "    'UUU': 'F', 'UCU': 'S', 'UAU': 'Y', 'UGU': 'C',  # UxU\n",
        "    'UUC': 'F', 'UCC': 'S', 'UAC': 'Y', 'UGC': 'C',  # UxC\n",
        "    'UUA': 'L', 'UCA': 'S', 'UAA': '-', 'UGA': '-',  # UxA\n",
        "    'UUG': 'L', 'UCG': 'S', 'UAG': '-', 'UGG': 'W',  # UxG\n",
        "    # C\n",
        "    'CUU': 'L', 'CCU': 'P', 'CAU': 'H', 'CGU': 'R',  # CxU\n",
        "    'CUC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',  # CxC\n",
        "    'CUA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',  # CxA\n",
        "    'CUG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',  # CxG\n",
        "    # A\n",
        "    'AUU': 'I', 'ACU': 'T', 'AAU': 'N', 'AGU': 'S',  # AxU\n",
        "    'AUC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',  # AxC\n",
        "    'AUA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',  # AxA\n",
        "    'AUG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',  # AxG\n",
        "    # G\n",
        "    'GUU': 'V', 'GCU': 'A', 'GAU': 'D', 'GGU': 'G',  # GxU\n",
        "    'GUC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',  # GxC\n",
        "    'GUA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',  # GxA\n",
        "    'GUG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'  # GxG\n",
        "    }\n",
        "    dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
        "    allowedBases = {'A','C','G','T','U','N'}\n",
        "    aa = {'A', 'G', 'M', 'S', 'C', 'H', 'N', 'T', 'D', 'I', 'P', 'V', 'E', 'K', 'Q', 'W', 'F', 'L', 'R', 'Y'}\n",
        "    aaComp = {}\n",
        "    aminoAcid = []\n",
        "    aminoAcidString = ''\n",
        "    nucComp = {}\n",
        "    codonComp = {}\n",
        "\n",
        "    def __init__ (self,seq):\n",
        "        self.seq = self.stripSequence(seq)\n",
        "\n",
        "    #builds all neccessary properties\n",
        "    def buildNucleotide(self):\n",
        "        self._toAminoAcid()\n",
        "        self._aaComposition()\n",
        "        self._nucComposition()\n",
        "        self._codonComposition()\n",
        "\n",
        "    #adds a genome sequence to the NucParams\n",
        "    def addSequence (self, thisSequence):\n",
        "        self.seq += self.stripSequence(thisSequence)\n",
        "\n",
        "    #cleans a string of unwanted bases, returns uppercase genome string\n",
        "    def stripSequence(self, protein):\n",
        "        splitSeq = []\n",
        "        for char in protein:\n",
        "            charUp = char.upper()\n",
        "            if charUp in self.allowedBases:\n",
        "                splitSeq.append(charUp)\n",
        "\n",
        "        return ''.join(''.join(splitSeq).split()).upper()\n",
        "\n",
        "    #returns a dictionary with counts of the 20 amino acids\n",
        "    def _aaComposition(self):\n",
        "        for amino in ProteinParam.aa2mw.keys():\n",
        "            self.aaComp[amino] = self.aminoAcid.count(amino)\n",
        "\n",
        "    #converts sequence to amino acid\n",
        "    def _toAminoAcid(self):\n",
        "        aminoAcid = []\n",
        "        newAA =  [self.seq[i:i+3] for i in range(0, len(self.seq), 3)]\n",
        "        self.codons = newAA\n",
        "        isRna = -1\n",
        "        for codon in newAA:\n",
        "            if codon in self.rnaCodonTable:\n",
        "                aminoAcid.append(self.rnaCodonTable[codon])\n",
        "            elif codon in self.dnaCodonTable:\n",
        "                aminoAcid.append(self.dnaCodonTable[codon])\n",
        "\n",
        "        self.aminoAcid = aminoAcid\n",
        "        self.aminoAcidString = ''.join(aminoAcid ).upper()\n",
        "\n",
        "    #returns a dictionary with counts of valid nucleotides\n",
        "    def _nucComposition(self):\n",
        "        for base in self.allowedBases:\n",
        "            self.nucComp[base] = self.seq.count(base)\n",
        "\n",
        "    #returns a dictionary with counts of valid codons\n",
        "    def _codonComposition(self):\n",
        "        for amino in self.rnaCodonTable.keys():\n",
        "            if 'N' in amino: continue\n",
        "            self.codonComp[amino] = self.codons.count(amino)\n",
        "\n",
        "    #returns a dictionary with counts of the 20 amino acids\n",
        "    def aaComposition(self):\n",
        "        return self.aaComp\n",
        "\n",
        "    #converts sequence to amino acid\n",
        "    def toAminoAcid(self):\n",
        "        return self.aminoAcidString\n",
        "\n",
        "    #returns a dictionary with counts of valid nucleotides\n",
        "    def nucComposition(self):\n",
        "        return self.nucComp\n",
        "\n",
        "    #returns a dictionary with counts of valid codons\n",
        "    def codonComposition(self):\n",
        "        return self.codonComp\n",
        "\n",
        "    #returns integer count of sum of valid nucleotides\n",
        "    def nucCount(self):\n",
        "        return len(self.seq)\n",
        "\n",
        "\n",
        "\n",
        "class FastAreader :\n",
        "    '''\n",
        "    Class to provide reading of a file containing one or more FASTA\n",
        "    formatted sequences:\n",
        "    object instantiation:\n",
        "    FastAreader(<file name>):\n",
        "\n",
        "    object attributes:\n",
        "    fname: the initial file name\n",
        "\n",
        "    methods:\n",
        "    readFasta() : returns header and sequence as strings.\n",
        "    Author: David Bernick\n",
        "    Date: April 19, 2013\n",
        "    '''\n",
        "    def __init__ (self, fname):\n",
        "        '''contructor: saves attribute fname '''\n",
        "        self.fname = fname\n",
        "\n",
        "    def readFasta (self):\n",
        "        '''\n",
        "        using filename given in init, returns each included FastA record\n",
        "        as 2 strings - header and sequence.\n",
        "        whitespace is removed, no adjustment is made to sequence contents.\n",
        "        The initial '>' is removed from the header.\n",
        "        '''\n",
        "        header = ''\n",
        "        sequence = ''\n",
        "\n",
        "        with open(self.fname) as fileH:\n",
        "            # initialize return containers\n",
        "            header = ''\n",
        "            sequence = ''\n",
        "\n",
        "            # skip to first fasta header\n",
        "            line = fileH.readline()\n",
        "            while not line.startswith('>') :\n",
        "                line = fileH.readline()\n",
        "            header = line[1:].rstrip()\n",
        "\n",
        "            # header is saved, get the rest of the sequence\n",
        "            # up until the next header is found\n",
        "            # then yield the results and wait for the next call.\n",
        "            # next call will resume at the yield point\n",
        "            # which is where we have the next header\n",
        "            for line in fileH:\n",
        "                if line.startswith ('>'):\n",
        "                    yield header,sequence\n",
        "                    header = line[1:].rstrip()\n",
        "                    sequence = ''\n",
        "                else :\n",
        "                    sequence += ''.join(line.rstrip().split()).upper()\n",
        "        # final header and sequence will be seen with an end of file\n",
        "        # with clause will terminate, so we do the final yield of the data\n",
        "        yield header,sequence\n",
        "\n",
        "# presumed object instantiation and example usage"
      ],
      "metadata": {
        "id": "BM4IRf5lfvE-"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### GenomeAnalyzer"
      ],
      "metadata": {
        "id": "8rRNozLjf874"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "paxqdziJeIgb"
      },
      "outputs": [],
      "source": [
        "# from sequenceAnalysis import NucParams,FastAreader,ProteinParam\n",
        "\n",
        "\"\"\"\n",
        "This program outputs the sequence length, GC content, and relative codon frequency for a given file containing genome information\n",
        "\"\"\"\n",
        "\n",
        "HUNDRED = 100\n",
        "NUC_DIVISOR = 1000000\n",
        "\n",
        "class genomeCompare:\n",
        "\tdef __init__ (self, filenames=['testGenome.fa'], compare=False):\n",
        "\t\tif compare:\n",
        "\t\t\tself.compareAnalysis(filenames)\n",
        "\t\telse:\n",
        "\t\t\tfor fa in filenames:\n",
        "\t\t\t\tself.individualAnalysis(fa)\n",
        "\n",
        "\tdef compareAnalysis(self,filenames):\n",
        "\t\tassert(len(filenames) == 2)\n",
        "\t\tprint('{} {}'.format(filenames[0],filenames[1]))\n",
        "\t\tfirstReader = FastAreader (filenames[0])\n",
        "\t\tsecondReader = FastAreader(filenames[1])\n",
        "\n",
        "\t\tfirstLength, firstNucParams = self.sequenceLength(firstReader)\n",
        "\t\tsecondLength, secondNucParams = self.sequenceLength(secondReader)\n",
        "\n",
        "\t\tprint('1st sequence length = {:.2f}Mb ----- 2nd sequence length = {:.2f}Mb'.format(firstLength,secondLength))\n",
        "\t\tprint ('sequence difference = {:.2f}Mb'.format(firstLength-secondLength))\n",
        "\n",
        "\t\tfirstGC = self.gcContent(firstNucParams)\n",
        "\t\tsecondGC = self.gcContent(secondNucParams)\n",
        "\n",
        "\t\tprint('1st GC content = {:.1f} ----- 2nd GC content = {:.1f}%'.format(firstGC,secondGC))\n",
        "\t\tprint ('GC difference = {:.1f}'.format(firstGC-secondGC))\n",
        "\n",
        "\t\tfirstCodonCount = firstNucParams.codonComposition()\n",
        "\t\tfirstAaComp = firstNucParams.aaComposition()\n",
        "\t\tfirstNucComp = firstNucParams.nucComposition()\n",
        "\n",
        "\t\tsecondCodonCount = secondNucParams.codonComposition()\n",
        "\t\tsecondAaComp = secondNucParams.aaComposition()\n",
        "\t\tsecondNucComp = secondNucParams.nucComposition()\n",
        "\n",
        "\t\tfor firstCodon,secondCodon in zip(sorted(firstCodonCount),sorted(secondCodonCount)):\n",
        "\t\t\tfirstTotal = firstCodonCount[firstCodon]\n",
        "\t\t\tfirstAmino = firstNucParams.rnaCodonTable[firstCodon]\n",
        "\t\t\tfirstAminoCount = firstNucParams.aminoAcid.count(firstAmino)\n",
        "\n",
        "\t\t\tsecondTotal = secondCodonCount[secondCodon]\n",
        "\t\t\tsecondAmino = secondNucParams.rnaCodonTable[secondCodon]\n",
        "\t\t\tsecondAminoCount = secondNucParams.aminoAcid.count(secondAmino)\n",
        "\n",
        "\t\t\tfirstFinalTotal = firstTotal/firstAminoCount * HUNDRED\n",
        "\t\t\tsecondFinalTotal = secondTotal/secondAminoCount * HUNDRED\n",
        "\n",
        "\t\t\tprint('1st: {} : {} {:5.1f}% ({:6d}) ----- 2nd: {} : {} {:5.1f}% ({:6d}) ----- difference: {:5.1f}%'.format(firstCodon,firstAmino,firstFinalTotal,firstTotal,secondCodon,secondAmino,secondFinalTotal,secondTotal,firstFinalTotal-secondFinalTotal))\n",
        "\n",
        "\n",
        "\tdef individualAnalysis(self, fa):\n",
        "\t\tprint ('Reading from file {}'.format(fa))\n",
        "\t\tmyReader = FastAreader (fa)\n",
        "\t\tlength,nucParams = self.sequenceLength(myReader)\n",
        "\t\tprint('sequence length = {:.2f}Mb'.format(length))\n",
        "\n",
        "\t\t#calculate and print GC content\n",
        "\t\tgc = self.gcContent(nucParams)\n",
        "\t\tprint('GC content = {:.1f}%'.format(gc))\n",
        "\n",
        "\t\t#calculate and print relative codon usage\n",
        "\t\tcodonCount = nucParams.codonComposition()\n",
        "\t\taaComp = nucParams.aaComposition()\n",
        "\t\tnucComp = nucParams.nucComposition()\n",
        "\t\tfor codon in sorted(codonCount):\n",
        "\t\t\ttotal = codonCount[codon]\n",
        "\t\t\tamino = nucParams.rnaCodonTable[codon]\n",
        "\t\t\taminoCount = nucParams.aminoAcid.count(amino)\n",
        "\n",
        "\t\t\tfinalTotal = total/aminoCount * HUNDRED\n",
        "\t\t\tprint('{} : {} {:5.1f}% ({:6d})'.format(codon,amino,finalTotal,total))\n",
        "\n",
        "\n",
        "\tdef gcContent(self,nucParams):\n",
        "\t\tnucComp = nucParams.nucComposition()\n",
        "\t\tgc = nucComp['G'] + nucComp['C']\n",
        "\t\tgc = gc/nucParams.nucCount() * HUNDRED\n",
        "\t\treturn gc\n",
        "\n",
        "\tdef sequenceLength(self,myReader):\n",
        "\n",
        "\t\tnucParams = NucParams('')\n",
        "\t\tfor head, seq in myReader.readFasta() :\n",
        "\t\t\tnucParams.addSequence(seq)\n",
        "\t\tnucParams.buildNucleotide()\n",
        "\n",
        "\t\tlength = nucParams.nucCount()/NUC_DIVISOR\n",
        "\t\treturn length, nucParams"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "genomeCompare(filenames=['testGenome.fa','haloVolc1_1-genes.fa'],compare=True)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DADZa-t2em_E",
        "outputId": "61b73764-f746-450f-876d-9e3f7d232a2b"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "testGenome.fa haloVolc1_1-genes.fa\n",
            "1st sequence length = 2.21Mb ----- 2nd sequence length = 3.54Mb\n",
            "sequence difference = -1.33Mb\n",
            "1st GC content = 106.3 ----- 2nd GC content = 66.3%\n",
            "GC difference = 40.0\n",
            "1st: AAA : K  18.4% (  7170) ----- 2nd: AAA : K  30.2% (  7170) ----- difference: -11.8%\n",
            "1st: AAC : N 146.7% ( 25502) ----- 2nd: AAC : N  90.1% ( 25502) ----- difference:  56.6%\n",
            "1st: AAG : K  42.5% ( 16582) ----- 2nd: AAG : K  69.8% ( 16582) ----- difference: -27.3%\n",
            "1st: AAU : N   0.0% (     0) ----- 2nd: AAU : N   0.0% (     0) ----- difference:   0.0%\n",
            "1st: ACA : T  11.4% (  3673) ----- 2nd: ACA : T   5.0% (  3673) ----- difference:   6.4%\n",
            "1st: ACC : T  98.2% ( 31675) ----- 2nd: ACC : T  43.4% ( 31675) ----- difference:  54.8%\n",
            "1st: ACG : T 105.7% ( 34092) ----- 2nd: ACG : T  46.7% ( 34092) ----- difference:  59.0%\n",
            "1st: ACU : T   0.0% (     0) ----- 2nd: ACU : T   0.0% (     0) ----- difference:   0.0%\n",
            "1st: AGA : R   4.9% (  2421) ----- 2nd: AGA : R   3.0% (  2421) ----- difference:   1.9%\n",
            "1st: AGC : S  42.7% ( 15288) ----- 2nd: AGC : S  22.0% ( 15288) ----- difference:  20.6%\n",
            "1st: AGG : R   3.1% (  1564) ----- 2nd: AGG : R   2.0% (  1564) ----- difference:   1.2%\n",
            "1st: AGU : S   0.0% (     0) ----- 2nd: AGU : S   0.0% (     0) ----- difference:   0.0%\n",
            "1st: AUA : I   0.0% (     0) ----- 2nd: AUA : I   0.0% (     0) ----- difference:   0.0%\n",
            "1st: AUC : I   0.0% (     0) ----- 2nd: AUC : I   0.0% (     0) ----- difference:   0.0%\n",
            "1st: AUG : M   0.0% (     0) ----- 2nd: AUG : M   0.0% (     0) ----- difference:   0.0%\n",
            "1st: AUU : I   0.0% (     0) ----- 2nd: AUU : I   0.0% (     0) ----- difference:   0.0%\n",
            "1st: CAA : Q  37.8% (  5779) ----- 2nd: CAA : Q  20.6% (  5779) ----- difference:  17.2%\n",
            "1st: CAC : H 186.6% ( 21315) ----- 2nd: CAC : H  90.9% ( 21315) ----- difference:  95.7%\n",
            "1st: CAG : Q 145.9% ( 22298) ----- 2nd: CAG : Q  79.4% ( 22298) ----- difference:  66.5%\n",
            "1st: CAU : H   0.0% (     0) ----- 2nd: CAU : H   0.0% (     0) ----- difference:   0.0%\n",
            "1st: CCA : P   8.2% (  3078) ----- 2nd: CCA : P   5.7% (  3078) ----- difference:   2.6%\n",
            "1st: CCC : P  55.3% ( 20681) ----- 2nd: CCC : P  38.1% ( 20681) ----- difference:  17.1%\n",
            "1st: CCG : P  77.1% ( 28825) ----- 2nd: CCG : P  53.2% ( 28825) ----- difference:  23.9%\n",
            "1st: CCU : P   0.0% (     0) ----- 2nd: CCU : P   0.0% (     0) ----- difference:   0.0%\n",
            "1st: CGA : R  22.3% ( 11054) ----- 2nd: CGA : R  13.8% ( 11054) ----- difference:   8.4%\n",
            "1st: CGC : R  78.7% ( 39088) ----- 2nd: CGC : R  48.8% ( 39088) ----- difference:  29.9%\n",
            "1st: CGG : R  42.2% ( 20971) ----- 2nd: CGG : R  26.2% ( 20971) ----- difference:  16.0%\n",
            "1st: CGU : R   0.0% (     0) ----- 2nd: CGU : R   0.0% (     0) ----- difference:   0.0%\n",
            "1st: CUA : L   0.0% (     0) ----- 2nd: CUA : L   0.0% (     0) ----- difference:   0.0%\n",
            "1st: CUC : L   0.0% (     0) ----- 2nd: CUC : L   0.0% (     0) ----- difference:   0.0%\n",
            "1st: CUG : L   0.0% (     0) ----- 2nd: CUG : L   0.0% (     0) ----- difference:   0.0%\n",
            "1st: CUU : L   0.0% (     0) ----- 2nd: CUU : L   0.0% (     0) ----- difference:   0.0%\n",
            "1st: GAA : E  60.6% ( 30617) ----- 2nd: GAA : E  32.8% ( 30617) ----- difference:  27.7%\n",
            "1st: GAC : D 281.9% ( 90365) ----- 2nd: GAC : D  92.1% ( 90365) ----- difference: 189.8%\n",
            "1st: GAG : E 123.9% ( 62633) ----- 2nd: GAG : E  67.2% ( 62633) ----- difference:  56.7%\n",
            "1st: GAU : D   0.0% (     0) ----- 2nd: GAU : D   0.0% (     0) ----- difference:   0.0%\n",
            "1st: GCA : A  10.2% (  7681) ----- 2nd: GCA : A   6.0% (  7681) ----- difference:   4.2%\n",
            "1st: GCC : A  73.3% ( 55219) ----- 2nd: GCC : A  42.9% ( 55219) ----- difference:  30.4%\n",
            "1st: GCG : A  79.6% ( 59953) ----- 2nd: GCG : A  46.6% ( 59953) ----- difference:  33.0%\n",
            "1st: GCU : A   0.0% (     0) ----- 2nd: GCU : A   0.0% (     0) ----- difference:   0.0%\n",
            "1st: GGA : G  15.8% (  8992) ----- 2nd: GGA : G   9.0% (  8992) ----- difference:   6.8%\n",
            "1st: GGC : G 106.3% ( 60394) ----- 2nd: GGC : G  60.7% ( 60394) ----- difference:  45.6%\n",
            "1st: GGG : G  33.8% ( 19173) ----- 2nd: GGG : G  19.3% ( 19173) ----- difference:  14.5%\n",
            "1st: GGU : G   0.0% (     0) ----- 2nd: GGU : G   0.0% (     0) ----- difference:   0.0%\n",
            "1st: GUA : V   0.0% (     0) ----- 2nd: GUA : V   0.0% (     0) ----- difference:   0.0%\n",
            "1st: GUC : V   0.0% (     0) ----- 2nd: GUC : V   0.0% (     0) ----- difference:   0.0%\n",
            "1st: GUG : V   0.0% (     0) ----- 2nd: GUG : V   0.0% (     0) ----- difference:   0.0%\n",
            "1st: GUU : V   0.0% (     0) ----- 2nd: GUU : V   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UAA : -   0.0% (     0) ----- 2nd: UAA : -   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UAC : Y   0.0% (     0) ----- 2nd: UAC : Y   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UAG : -   0.0% (     0) ----- 2nd: UAG : -   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UAU : Y   0.0% (     0) ----- 2nd: UAU : Y   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UCA : S   0.0% (     0) ----- 2nd: UCA : S   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UCC : S   0.0% (     0) ----- 2nd: UCC : S   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UCG : S   0.0% (     0) ----- 2nd: UCG : S   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UCU : S   0.0% (     0) ----- 2nd: UCU : S   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UGA : -   0.0% (     0) ----- 2nd: UGA : -   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UGC : C   0.0% (     0) ----- 2nd: UGC : C   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UGG : W   0.0% (     0) ----- 2nd: UGG : W   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UGU : C   0.0% (     0) ----- 2nd: UGU : C   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UUA : L   0.0% (     0) ----- 2nd: UUA : L   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UUC : F   0.0% (     0) ----- 2nd: UUC : F   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UUG : L   0.0% (     0) ----- 2nd: UUG : L   0.0% (     0) ----- difference:   0.0%\n",
            "1st: UUU : F   0.0% (     0) ----- 2nd: UUU : F   0.0% (     0) ----- difference:   0.0%\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<__main__.genomeCompare at 0x7a1752138f40>"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    }
  ]
}